metrics/metrics.go

```diff
@@ -84,327 +140,56 @@ type Options struct {
-	o.EnableRouteBackendMetrics = true
-
-	return o
+func init() {
+	Void = NewVoid()
+	Default = Void
 }
 
-func New(o Options) *Metrics {
-	o = applyCompatibilityDefaults(o)
-
-	m := &Metrics{}
-	m.reg = metrics.NewRegistry()
-
-	var createSample func() metrics.Sample
-	if o.UseExpDecaySample {
-		createSample = newExpDecaySample
-	} else {
-		createSample = newUniformSample
-	}
-	m.createTimer = func() metrics.Timer { return createTimer(createSample()) }
+// NewDefaultHandler returns a default metrics handler.
+func NewDefaultHandler(o Options) http.Handler {
+	var m Metrics
 
-	m.createCounter = metrics.NewCounter
-	m.options = o
-
-	if o.EnableDebugGcMetrics {
-		metrics.RegisterDebugGCStats(m.reg)
-		go metrics.CaptureDebugGCStats(m.reg, statsRefreshDuration)
-	}
-
-	if o.EnableRuntimeMetrics {
-		metrics.RegisterRuntimeMemStats(m.reg)
-		go metrics.CaptureRuntimeMemStats(m.reg, statsRefreshDuration)
+	switch o.Format {
+	case PrometheusKind:
+		m = NewPrometheus(o)
+	default:
+		// CodaHale is the default backend always.
+		m = NewCodaHale(o)
 	}
 
-	return m
-}
-
-func NewVoid() *Metrics {
-	m := &Metrics{}
-	m.reg = metrics.NewRegistry()
-	m.createTimer = func() metrics.Timer { return metrics.NilTimer{} }
-	m.createCounter = func() metrics.Counter { return metrics.NilCounter{} }
-	return m
-}
-
-func init() {
-	Void = NewVoid()
-	Default = Void
+	return NewHandler(o, m)
 }
 
 // NewHandler returns a collection of metrics handlers.
-func NewHandler(o Options) http.Handler {
-	Default = New(o)
+func NewHandler(o Options, m Metrics) http.Handler {
 
-	handler := &metricsHandler{registry: Default.reg, options: o}
+	mux := http.NewServeMux()
 	if o.EnableProfile {
-		mux := http.NewServeMux()
 		mux.Handle("/debug/pprof/", http.HandlerFunc(pprof.Index))
 		mux.Handle("/debug/pprof/cmdline", http.HandlerFunc(pprof.Cmdline))
 		mux.Handle("/debug/pprof/profile", http.HandlerFunc(pprof.Profile))
 		mux.Handle("/debug/pprof/symbol", http.HandlerFunc(pprof.Symbol))
 		mux.Handle("/debug/pprof/trace", http.HandlerFunc(pprof.Trace))
-		handler.profile = mux
-	}
-
-	return handler
-}
-
-func newUniformSample() metrics.Sample {
-	return metrics.NewUniformSample(defaultUniformReservoirSize)
-}
-
-func newExpDecaySample() metrics.Sample {
-	return metrics.NewExpDecaySample(defaultExpDecayReservoirSize, defaultExpDecayAlpha)
-}
-
-func createTimer(sample metrics.Sample) metrics.Timer {
-	return metrics.NewCustomTimer(metrics.NewHistogram(sample), metrics.NewMeter())
-}
-
-func (m *Metrics) getTimer(key string) metrics.Timer {
-	return m.reg.GetOrRegister(key, m.createTimer).(metrics.Timer)
-}
-
-func (m *Metrics) updateTimer(key string, d time.Duration) {
-	if t := m.getTimer(key); t != nil {
-		t.Update(d)
 	}
-}
 
-func (m *Metrics) MeasureSince(key string, start time.Time) {
-	m.measureSince(key, start)
-}
+	// Root path should return 404.
+	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusBadRequest)
```

Comment:
i'd actually agree with the comment above, it's more like a 404 (http.StatusNotFound) than a 400